usr/man/man/usr/share/man/man1/comm.1
'\" te
.\" Copyright (c) 2009, 2011, Oracle and/or its affiliates. All rights reserved.
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited. All Rights Reserved.
.\" Portions Copyright (c) 1982-2007 AT&T Knowledge Ventures
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for permission to reproduce portions of its copyrighted documentation. Original documentation from The Open Group can be obtained online at http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group, have given us permission to reprint portions of their documentation. In the following statement, the phrase "this text" refers to portions of the system documentation. Portions of this text are reprinted and reproduced in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1, 2004 Edition, Standard for Information Technology -- Portable Operating System Interface (POSIX), The Open Group Base Specifications Issue 6, Copyright (C) 2001-2004 by the Institute of Electrical and Electronics Engineers, Inc and The Open Group. In the event of any discrepancy between these versions and the original IEEE and The Open Group Standard, the original IEEE and The Open Group Standard is the referee document. The original Standard can be obtained online at http://www.opengroup.org/unix/online.html.  This notice shall appear on any product containing this material.
.TH comm 1 "26 Jul 2011" "SunOS 5.11" "User Commands"
.SH NAME
comm \- select or reject lines common to two files
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/comm\fR [\fB-\fIoptions\fR\fR] \fIfile1\fR \fIfile2\fR
.fi

.SH DESCRIPTION
.sp
.LP
\fBcomm\fR reads two files \fIfile1\fR and \fIfile2\fR which should be ordered in the collating sequence of the current locale, and produces three text columns as output:
.sp
.ne 2
.mk
.na
\fB1\fR
.ad
.RS 5n
.rt  
Lines only in \fIfile1\fR.
.RE

.sp
.ne 2
.mk
.na
\fB2\fR
.ad
.RS 5n
.rt  
Lines only in \fIfile2\fR.
.RE

.sp
.ne 2
.mk
.na
\fB3\fR
.ad
.RS 5n
.rt  
Lines in both files.
.RE

.sp
.LP
If lines in either file are not ordered according to the collating sequence of the current locale, the results are not specified.
.sp
.LP
If either \fIfile1\fR or \fIfile2\fR is \fB-\fR, \fBcomm\fR uses standard input starting at the current location.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB-1\fR\fR
.ad
.RS 6n
.rt  
Suppresses the output column of lines unique to \fIfile1\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-2\fR\fR
.ad
.RS 6n
.rt  
Suppresses the output column of lines unique to \fIfile2\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-3\fR\fR
.ad
.RS 6n
.rt  
Suppresses the output column of lines duplicated in \fIfile1\fR and \fIfile2\fR.
.RE

.SH OPERANDS
.sp
.LP
The following operands are supported:
.sp
.ne 2
.mk
.na
\fB\fIfile1\fR\fR
.ad
.RS 9n
.rt  
A path name of the first file to be compared. If \fIfile1\fR is \fB\(mi\fR, the standard input is used.
.RE

.sp
.ne 2
.mk
.na
\fB\fIfile2\fR\fR
.ad
.RS 9n
.rt  
A path name of the second file to be compared. If \fIfile2\fR is \fB\(mi\fR, the standard input is used.
.RE

.SH USAGE
.sp
.LP
See \fBlargefile\fR(5) for the description of the behavior of \fBcomm\fR when encountering files greater than or equal to 2 Gbyte (2^31 bytes).
.SH EXAMPLES
.LP
\fBExample 1 \fRPrinting a list of utilities specified by files
.sp
.LP
If \fIfile1\fR, \fIfile2\fR, and \fIfile3\fR each contain a sorted list of utilities, the command

.sp
.in +2
.nf
example% \fBcomm -23 file1 file2 | comm -23 - file3\fR 
.fi
.in -2
.sp

.sp
.LP
prints a list of utilities in \fIfile1\fR not specified by either of the other files. The entry:

.sp
.in +2
.nf
example% \fBcomm -12 file1 file2 | comm -12 - file3\fR
.fi
.in -2
.sp

.sp
.LP
prints a list of utilities specified by all three files. And the entry:

.sp
.in +2
.nf
example% \fBcomm -12 file2 file3 | comm -23 -file1\fR
.fi
.in -2
.sp

.sp
.LP
prints a list of utilities specified by both \fIfile2\fR and \fIfile3\fR, but not specified in \fIfile1\fR.

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(5) for descriptions of the following environment variables that affect the execution of \fBcomm\fR: \fBLANG\fR, \fBLC_ALL\fR, \fBLC_COLLATE\fR, \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.mk
.na
\fB\fB0\fR\fR
.ad
.RS 6n
.rt  
All input files were successfully output as specified.
.RE

.sp
.ne 2
.mk
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
.rt  
An error occurred.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
_
CSIEnabled
_
Interface StabilityCommitted
_
StandardSee \fBstandards\fR(5).
.TE

.SH SEE ALSO
.sp
.LP
\fBcmp\fR(1), \fBdiff\fR(1), , \fBsort\fR(1), \fBuniq\fR(1), \fBattributes\fR(5), \fBenviron\fR(5), \fBlargefile\fR(5), \fBstandards\fR(5)


usr/src/cmd/comm/comm.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 1988, 2011, Oracle and/or its affiliates. All rights reserved.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*
 *	process common lines of two files
 */

#include 	<locale.h>
#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>

static int compare(char *, char *);
static int rd(FILE *, char *);
static FILE *openfil(char *);
static void copy(FILE *, char *, int);
static void usage(void);
static void wr(char *, int);

#define	LB	2050	/* P1003.2 minimum (2048) + 2 */

#define	RDTWO(ib1, lb1, ib2, lb2) \
	{ \
		if (rd(ib1, lb1) < 0) { \
			if (rd(ib2, lb2) < 0) \
				exit(0); \
			copy(ib2, lb2, 2); \
		} \
		if (rd(ib2, lb2) < 0) \
			copy(ib1, lb1, 1); \
	}

static int	one;
static int	two;
static int	three;

static char	ldr[3][3] = {"", "\t", "\t\t"};

static FILE	*ib1;
static FILE	*ib2;
static int	is_c_locale;

int
main(int argc, char **argv)
{
	int	l = 1;
	int	c;		/* used for getopt() */
	char	lb1[LB], lb2[LB], *collate;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)	/* Should be defined by cc -D */
#define	TEXT_DOMAIN "SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	if ((collate = setlocale(LC_COLLATE, NULL)) == NULL) {
		(void) fprintf(stderr,
		    gettext("Query of LC_COLLATE category failed\n"));
		exit(4);
	}

	is_c_locale = (strcmp("C", collate) == 0) ? 1 : 0;
	while ((c = getopt(argc, argv, "123")) != EOF)
		switch (c) {
		case '1':
			if (!one) {
				one = 1;
				ldr[1][0] = ldr[2][l--] = '\0';
			}
			break;
		case '2':
			if (!two) {
				two = 1;
				ldr[2][l--] = '\0';
			}
			break;
		case '3':
			three = 1;
			break;

		default:
			usage();
		}

	argc -= optind;
	argv  = &argv[optind];

	if (argc != 2)
		usage();
	ib1 = openfil(argv[0]);
	ib2 = openfil(argv[1]);
	RDTWO(ib1, lb1, ib2, lb2);
	for (;;) {
		switch (compare(lb1, lb2)) {
			case 0:
				wr(lb1, 3);
				RDTWO(ib1, lb1, ib2, lb2);
				continue;

			case 1:
				wr(lb1, 1);
				if (rd(ib1, lb1) < 0)
					copy(ib2, lb2, 2);
				continue;

			case 2:
				wr(lb2, 2);
				if (rd(ib2, lb2) < 0)
					copy(ib1, lb1, 1);
				continue;
			/*
			 * case "3" means lines are equal in collation,
			 * but not identical (not very likely)
			 */
			case 3:
				wr(lb1, 1);
				wr(lb2, 2);
				RDTWO(ib1, lb1, ib2, lb2);
				continue;
		}
	}
	/* NOTREACHED */
	return (0);
}

static int
rd(file, buf)
FILE *file;
char *buf;
{
	register int i, j;
	i = j = 0;
	while ((j = getc(file)) != EOF) {
		*buf = (char)j;
		if (*buf == '\n' || i > LB-2) {
			*buf = '\0';
			return (0);
		}
		i++;
		buf++;
	}
	return (-1);
}

static void
wr(str, n)
char *str;
int n;
{
	switch (n) {
		case 1:
			if (one)
				return;
			break;

		case 2:
			if (two)
				return;
			break;

		case 3:
			if (three)
				return;
	}
	(void) printf("%s%s\n", ldr[n-1], str);
}

static void
copy(ibuf, lbuf, n)
FILE *ibuf;
char *lbuf;
int n;
{
	do {
		wr(lbuf, n);
	} while (rd(ibuf, lbuf) >= 0);

	exit(0);
}

static int
compare(a, b)
char *a, *b;
{
	register char *ra, *rb;
	int ret;

	ra = a - 1;
	rb = b - 1;
	while (*++ra == *++rb)
		if (*ra == '\0')
			return (0);

	/* For "C" locale, just compare bytes */
	if (is_c_locale) {
		if (*ra < *rb)
			return (1);
		return (2);
	}
	/* For other locales, call locale-sensitive compare routine */
	else {
		ret = strcoll(a, b);
		return (ret == 0 ? 3 : (ret < 0 ? 1 : 2));
	}
}

static FILE *
openfil(s)
char *s;
{
	FILE *b;
	if (s[0] == '-' && s[1] == 0)
		b = stdin;
	else if ((b = fopen(s, "r")) == NULL) {
		(void) fprintf(stderr, "comm: ");
		perror(s);
		exit(2);
	}
	return (b);
}

static void
usage()
{
	(void) fprintf(stderr, gettext("usage: comm [-123] file1 file2\n"));
	exit(2);
}
